### ISSUE TITLE
`clap` should not be ignored if marked essential

### ISSUE THREAD

**Issue by leo-barnes** [AOMediaCodec/libavif#1177]
Chrome and Firefox are currently ignoring the `clap` item property due to the potential privacy concerns. I'm not sure if this is due to how libavif implements it or something that is done in their respective implementations.

If the `clap` property is marked as essential, I would argue that the correct behaviour according to the specs should really be to reject the files altogether. In other words, treat `clap` as any other unknown item property that is marked essential.

**Comment by joedrago** [comment]
I'd defer to you and @cconcolato on whether or not erroring out here is the best choice. That said, there is a codepath for such a thing here:

https://github.com/AOMediaCodec/libavif/blob/main/src/read.c#L2060

If we added `"clap"` to the `nonessentialTypes[]` array from the link above, libavif would behave as you're recommending (`clap` props marked essential would gracefully error out with a clear diagnostic error). It would be quite easy to change the behavior to this recommended path, should that end up being the final decision. 

Is there any snippet in the standard we could refer to in such a change that would corroborate the decision to add this?

**Comment by leo-barnes** [comment]
HEIF spec:
> 9.3 Item Properties Box
> ...
> Each property association may be marked as either essential or non-essential. **A reader shall not process an item that is associated with a property that is not recognized or not supported by the reader and that is marked as essential to the item**. A reader may ignore an associated item property that is marked non-essential to the item. 

It is perfectly fine for libavif to not support `clap`, but if it does, it should really be treated as an unknown/unsupported property. If such a property is marked essential, the file should really be rejected, since you're not applying a property that the file creator thought was needed to correctly decode the file.

**Comment by joedrago** [comment]
To be clear, libavif itself has full support for reading and interpreting/validating clap, but simply surfaces this transformation data to the caller (Chrome, in this case).

This library really can't legislate whether or not the caller chooses to ignore some subset of the attached metadata to an image (including color profiles, rotations, or other transformation data). I'm not arguing with you on whether or not something should be done, I just think that maybe this is more of a decision for the Chrome team than anything that should change in this library.

**Comment by joedrago** [comment]
I suppose in order for the caller to behave exactly the way you are requesting (assuming they want to continue to ignore clap), this library would have to surface that the transformation was flagged as essential. we probably have to come up with a clean way to surface that.

**Comment by leo-barnes** [comment]
> To be clear, libavif itself has full support for reading and interpreting/validating clap, but simply surfaces this transformation data to the caller (Chrome, in this case).
>
> This library really can't legislate whether or not the caller chooses to ignore some subset of the attached metadata to an image (including color profiles, rotations, or other transformation data). I'm not arguing with you on whether or not something should be done, I just think that maybe this is more of a decision for the Chrome team than anything that should change in this library.

Right. In that case I think libavif is doing the correct thing here and this issue can be closed.
I'll try to file issues for Chrome and Firefox. 

@wantehchang 
Where would I file an issue for Chrome? (If you know where to file an issue for Firefox that would be very helpful as well :) )

> I suppose in order for the caller to behave exactly the way you are requesting (assuming they want to continue to ignore clap), this library would have to surface that the transformation was flagged as essential. we probably have to come up with a clean way to surface that.

I guess this would be ideal, but I can't really see when you would ever have `clap` that is not marked essential. Same goes for rotation or mirroring. How should a reader interpret a file that has "optional" rotation?

**Comment by y-guyon** [comment]
> Where would I file an issue for Chrome?

The platform would be monorail: https://bugs.chromium.org/p/chromium/issues/list
There is already one bug on the topic, maybe no need for another: https://bugs.chromium.org/p/chromium/issues/detail?id=1167422
